apiserver/restrict_newer_client_test.go

```diff
@@ -56,67 +58,101 @@ func (r *restrictNewerClientSuite) TestRecentClientAllowedAll(c *gc.C) {
 	c.Assert(err, jc.Satisfies, params.IsIncompatibleClientError)
 	c.Assert(caller, gc.IsNil)
 }
 
 func (r *restrictNewerClientSuite) TestReallyNewClientDisallowedMethod(c *gc.C) {
-	r.olderVersion.Major = jujuversion.Current.Major + 2
-	root := apiserver.TestingUpgradeOrMigrationOnlyRoot(true, r.olderVersion)
-	caller, err := root.FindMethod("Client", clientFacadeVersion, "FullStatus")
+	r.callerVersion.Major = jujuversion.Current.Major + 2
+	root := apiserver.TestingUpgradeOrMigrationOnlyRoot(true, r.callerVersion)
+	caller, err := root.FindMethod("Client", 3, "FullStatus")
 	c.Assert(err, jc.Satisfies, params.IsIncompatibleClientError)
 	c.Assert(caller, gc.IsNil)
 }
 
 func (r *restrictNewerClientSuite) TestAlwaysDisallowedMethod(c *gc.C) {
-	root := apiserver.TestingUpgradeOrMigrationOnlyRoot(true, r.olderVersion)
-	caller, err := root.FindMethod("Client", clientFacadeVersion, "ModelSet")
+	r.callerVersion.Major = jujuversion.Current.Major - 1
+	root := apiserver.TestingUpgradeOrMigrationOnlyRoot(true, r.callerVersion)
+	caller, err := root.FindMethod("ModelConfig", 3, "ModelSet")
 	c.Assert(err, jc.Satisfies, params.IsIncompatibleClientError)
 	c.Assert(caller, gc.IsNil)
 }
 
-func (r *restrictNewerClientSuite) TestAgentAllowedMethod(c *gc.C) {
-	r.olderVersion.Major = 2
-	r.olderVersion.Minor = 9
-	root := apiserver.TestingUpgradeOrMigrationOnlyRoot(false, r.olderVersion)
-	checkAllowed := func(facade, method string, version int) {
-		caller, err := root.FindMethod(facade, version, method)
+func (r *restrictNewerClientSuite) TestWhitelistedClient(c *gc.C) {
+	r.assertWhitelistedClient(c, "2.9.32", false)
+	r.assertWhitelistedClient(c, "2.9.33", true)
+}
+
+func (r *restrictNewerClientSuite) assertWhitelistedClient(c *gc.C, serverVers string, allowed bool) {
+	r.PatchValue(&jujuversion.Current, version.MustParse(serverVers))
+	r.callerVersion = version.MustParse("3.0.0")
+	root := apiserver.TestingUpgradeOrMigrationOnlyRoot(true, r.callerVersion)
+	caller, err := root.FindMethod("ModelConfig", 3, "ModelSet")
+	if allowed {
 		c.Check(err, jc.ErrorIsNil)
 		c.Check(caller, gc.NotNil)
+	} else {
+		c.Check(err, gc.NotNil)
+		c.Check(caller, gc.IsNil)
 	}
-	checkAllowed("Uniter", "CurrentModel", 18)
 }
 
-func (r *restrictNewerClientSuite) TestReallyOldAgentDisallowedMethod(c *gc.C) {
-	r.olderVersion.Minor = 0
-	root := apiserver.TestingUpgradeOrMigrationOnlyRoot(true, r.olderVersion)
+func (r *restrictNewerClientSuite) TestAgentMethod(c *gc.C) {
+	r.PatchValue(&jujuversion.Current, version.MustParse("3.0.0"))
+	r.assertAgentMethod(c, "2.9.32", true)
+	r.assertAgentMethod(c, "2.9.33", true)
+	r.assertAgentMethod(c, "2.9.31", false)
+}
+
+func (r *restrictNewerClientSuite) assertAgentMethod(c *gc.C, agentVers string, allowed bool) {
+	r.callerVersion = version.MustParse(agentVers)
+	root := apiserver.TestingUpgradeOrMigrationOnlyRoot(false, r.callerVersion)
 	caller, err := root.FindMethod("Uniter", 15, "CurrentModel")
```

Comment:
@jack-w-shaw It passes if you use version 18 here, which is the only one now present in 3.0

Reply:
Thanks!